### PR TITLE
Port Memory remote document cache to C++

### PR DESCRIPTION
**Pull request by gsoltis**

### CODE REVIEW

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm [hunk 1]
```diff
@@ -33,106 +35,143 @@
...
     const DocumentKey &key = it->first;
     if (!query.path.IsPrefixOf(key.path())) {
       break;
     }
-    FSTMaybeDocument *maybeDoc = nil;
-    auto found = _docs.find(key);
-    if (found != _docs.end()) {
-      maybeDoc = found->second;
-    }
+    FSTMaybeDocument *maybeDoc = it->second;
```
Comment (var-const) [inline Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm]: Thanks for fixing this, I was overly literal when translating from Objective-C.

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm [hunk 2]
```diff
@@ -33,106 +35,143 @@
...
   }
-
-  return result;
+  return results;
 }
 
-- (std::vector<DocumentKey>)removeOrphanedDocuments:
-                                (FSTMemoryLRUReferenceDelegate *)referenceDelegate
-                              throughSequenceNumber:(ListenSequenceNumber)upperBound {
+std::vector<model::DocumentKey> MemoryRemoteDocumentCache::RemoveOrphanedDocuments(
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm]: Fixed.

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm [hunk 3]
```diff
@@ -33,106 +35,143 @@
...
-- (nullable FSTMaybeDocument *)entryForKey:(const DocumentKey &)key {
-  auto found = _docs.find(key);
-  return found != _docs.end() ? found->second : nil;
+FSTMaybeDocument *_Nullable MemoryRemoteDocumentCache::Find(const DocumentKey &key) {
+  auto found = docs_.find(key);
+  return found != docs_.end() ? found->second : nil;
 }
 
-- (MaybeDocumentMap)entriesForKeys:(const DocumentKeySet &)keys {
+model::MaybeDocumentMap MemoryRemoteDocumentCache::FindAll(const model::DocumentKeySet &keys) {
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm]: Fixed.

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm [hunk 4]
```diff
@@ -33,106 +35,143 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@implementation FSTMemoryRemoteDocumentCache {
+  MemoryRemoteDocumentCache _cache;
+}
+
+- (instancetype)init {
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply Firestore/Source/Local/FSTMemoryRemoteDocumentCache.mm]: You're right. I removed it.

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.h [hunk 5]
```diff
@@ -37,4 +39,36 @@ NS_ASSUME_NONNULL_BEGIN
 
 @end
 
+namespace firebase {
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply Firestore/Source/Local/FSTMemoryRemoteDocumentCache.h]: Ok, done.

File: Firestore/Source/Local/FSTMemoryRemoteDocumentCache.h [hunk 6]
```diff
@@ -37,4 +39,36 @@ NS_ASSUME_NONNULL_BEGIN
...
+namespace firestore {
+namespace local {
+
+class MemoryRemoteDocumentCache {
+ public:
+  MemoryRemoteDocumentCache();
+
+  void AddEntry(FSTMaybeDocument *document);
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Source/Local/FSTMemoryRemoteDocumentCache.h]: Ah, sorry, I missed that the comments were on the interface.

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  FSTMaybeDocument *_Nullable Get(const model::DocumentKey &key);
+
+  model::MaybeDocumentMap GetAll(const model::DocumentKeySet &keys);
+
+  model::DocumentMap GetMatchingDocuments(FSTQuery *query);
+
+  std::vector<model::DocumentKey> RemoveOrphanedDocuments(
+          FSTMemoryLRUReferenceDelegate *reference_delegate, model::ListenSequenceNumber upper_bound);
+
+  size_t CalculateByteSize(FSTLocalSerializer *serializer);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h]: Nit: this function can be `const`.

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class MemoryRemoteDocumentCache {
+ public:
+  void AddEntry(FSTMaybeDocument *document);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h]: Optional: maybe group `AddEntry` together with `RemoveEntry`, and `Get` with `GetAll`/maybe also `GetMatchingDocuments`? (I mean, remove empty lines between them).

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.mm [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+namespace firestore {
+namespace local {
+
+namespace {
+/**
+ * Returns an estimate of the number of bytes used to store the given
+ * document key in memory. This is only an estimate and includes the size
+ * of the segments of the path, but not any object overhead or path separators.
+ */
+static size_t DocumentKeyByteSize(const DocumentKey &key) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.mm]: Nit: now that this function is in an anonymous namespace, `static` is redundant.

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h (hunk 7 above)
Comment (gsoltis) [reply Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h]: As discussed, there will eventually be a leveldb version that will access non-const methods.

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h (hunk 8 above)
Comment (gsoltis) [reply Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h]: Done.

File: Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.mm (hunk 9 above)
Comment (gsoltis) [reply Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.mm]: Done.

**Pull request merged**